src/libraries/System.Net.NameResolution/src/System/Net/NameResolutionTelemetry.cs

```diff
@@ -0,0 +1,214 @@
+
+                _lookupsDuration!.WriteMetric(duration);
+
+                WriteEvent(
+                    successful ? ResolutionSuccessEventId : ResolutionFailureEventId,
+                    FormatIPAddressNullTerminated(address, stackalloc char[MaxIPFormattedLength]),
+                    duration);
+            }
+        }
+
+
+        [Event(ResolutionSuccessEventId, Level = EventLevel.Informational)]
+        private void ResolutionSuccess(string hostNameOrAddress, double duration) => WriteEvent(ResolutionSuccessEventId, hostNameOrAddress, duration);
+
+        [Event(ResolutionFailureEventId, Level = EventLevel.Informational)]
+        private void ResolutionFailure(string hostNameOrAddress, double duration) => WriteEvent(ResolutionFailureEventId, hostNameOrAddress, duration);
+
+
+        [NonEvent]
+        private static Span<char> FormatIPAddressNullTerminated(IPAddress address, Span<char> destination)
+        {
+            Debug.Assert(address != null);
+
+            bool success = address.TryFormat(destination, out int charsWritten);
+            Debug.Assert(success);
+
+            Debug.Assert(charsWritten < destination.Length);
+            destination[charsWritten] = '\0';
+
+            return destination.Slice(0, charsWritten + 1);
+        }
+
+
+        [NonEvent]
+        private unsafe void WriteEvent(int eventId, Span<char> arg1)
+        {
+            Debug.Assert(!arg1.IsEmpty && arg1[^1] == '\0', "Expecting a null-terminated ROS<char>");
+
+            if (IsEnabled())
+            {
+                fixed (char* arg1Ptr = &MemoryMarshal.GetReference(arg1))
+                {
+                    EventData descr = new EventData
+                    {
+                        DataPointer = (IntPtr)(arg1Ptr),
+                        Size = arg1.Length * sizeof(char)
+                    };
+
+                    WriteEventCore(eventId, eventDataCount: 1, &descr);
+                }
+            }
+        }
+
+        [NonEvent]
+        private unsafe void WriteEvent(int eventId, string? arg1, double arg2)
+        {
+            if (IsEnabled())
+            {
+                arg1 ??= "";
+
+                fixed (char* arg1Ptr = arg1)
+                {
+                    const int NumEventDatas = 2;
+                    EventData* descrs = stackalloc EventData[NumEventDatas];
+
+                    descrs[0] = new EventData
+                    {
+                        DataPointer = (IntPtr)(arg1Ptr),
+                        Size = (arg1.Length + 1) * sizeof(char)
+                    };
+                    descrs[1] = new EventData
+                    {
+                        DataPointer = (IntPtr)(&arg2),
+                        Size = sizeof(double)
+                    };
+
+                    WriteEventCore(eventId, NumEventDatas, descrs);
+                }
+            }
+        }
+
+        [NonEvent]
+        private unsafe void WriteEvent(int eventId, Span<char> arg1, double arg2)
```

Comment:
Do we need both the above overload and this one?  This one can be made to take a `ReadOnlySpan<char>` instead of a `Span<char>`, and then delete the other one.

Reply:
I kept both to differentiate them based on null-termination handling.

I can reuse the ROS<char> one and debug assert that the ROS<char> is null terminated _after_ its end

Reply:
Why does null termination matter?  The exact length is passed down to the WriteEventCore.

Reply:
> The exact length is passed down to the WriteEventCore

For the Span overload yes, as it's ensured that the last character is null. For the string overload `Length + 1` is passed.

Reply:
@brianrob, given EventData includes the Size, do you know why strings need to be null terminated?

Reply:
The reason is mostly separation of concerns.  The ETW runtime doesn't know the type of the data or the encoding - it just gets a blob and a size and uses the size to write out the appropriate amount of the blob to the event stream.  At this point, the size is thrown away (it is not persisted into the etl).

On the other side, the consumer knows how to decode the event payload.  NULL-terminated strings don't have a size associated with them in their written format - the consumer will just read until it hits a NULL.  There is another string encoding called counted string where the count is pre-pended to the string, and then you don't need the NULL char.  If `EventSource` wrote counted strings, it would be responsible for prepending the count and then specifying the pointer to the count+string as the value assigned to `DataPointer`.  Then `Size` would be the combination of the two.  But again, `Size` would be thrown away and not persisted by the ETW runtime, but the consumer would know it is a counted string and so it would not depend on a NULL char, but would require a count to know how big the string is.

Reply:
In that case, shouldn't there at least be a Debug asset in EventSource to catch null chars in string payloads?

Reply:
Unfortunately we can't really do that because we ship release binaries, so by the time you get into `EventSource` code, the asserts have all been stripped out.  The exception to this is for framework eventsources, but that's a pretty small exception.

This is one of the classic issues that has been discussed time and time again, and so far we've not found a nice and pay-for-play way to address this.  It's considered by design that if you pass a string with embedded nulls, you're going to have incorrect decoding behavior in the consumer.  That's the case for both ETW and EventPipe.  I don't know enough about the serialization mechanism to be able to say for sure with LTTng, and I've not tested it.

Reply:
Considering that the overloads have different assumptions about the input (Span includes the null character, string has one implicitly at the end), I believe these should remain as separate overloads.